code/go/0chain.net/chaincore/chain/worker.go

```diff
@@ -340,12 +352,39 @@ func (c *Chain) finalizeBlockProcess(ctx context.Context, fb *block.Block, bsh B
 		var err = c.repairChain(ctx, fb, bsh.SaveMagicBlock())
 		if err != nil {
 			Logger.Error("finalize block - repairing MB chain", zap.Error(err))
-			return
+			return fmt.Errorf("repair chain failed: %v", err)
 		}
 	}
 
+	if isSharder {
+		// get previous finalized block
+		pr := c.GetRound(fb.Round - 1)
+		if pr == nil {
+			Logger.Error("finalize block - previous round not found",
+				zap.Int64("round", fb.Round))
+			return errors.New("previous round is missing")
+		}
+
+		prevBlockHash := pr.GetBlockHash()
+		if prevBlockHash == "" {
```

Comment:
probably, I'm missing something. sometimes previous round can't be finalize, since it has more than one notarizations, this round will be finalized by chain, when it's siblings will have singlefinalized block on a round. is it different for sharders?

Reply:
We are finalizing blocks in `frchain` from LFB+1, it's normal for miners that the previous round is not finalized as they could be restarted and come up with LFB acquired from remote. But for sharders, they all have full blocks till LFB, the frchain is start from LFB+1, and blocks in it are all chained, also we finalize blocks one by one, so it's previous round is actually finalized. If its siblings have single finalized block, but since they are not in the frchain, that means the sibling singlefinalized block does not have enough successors. 

BTW, if I don't do this checking, the previous logic will finalize the block anyway, which means we will finalize the block even it's previous round is not finalized. Apparently this is not what we want, if this happens, the event db will be messed up.

Reply:
> which means we will finalize the block even it's previous round is not finalized. 
it's normal for our protocol, previous round can be not finalized, 
so we can have b0<-b1<-b2<-b3
imagine we want to finalize b3 now, b0 can be finalized and b1 and b2 can be not finalized, simply because they have other notarized blocks on this round due to leaders equivocations. 
if b3 is the only on his round (is in the lock set, b4 is build upon it and finalized) then b3 can be finalized and at this moment b2 and b3 will be finalized, not earlier.

this branches of notarized can be build by a sequence of equivocating leaders, they should cooperate. we have this byzantine test (not implemented yet), in general such chains won't be too long, since it is not likely to have several dishonest leaders in a row, but two leaders can cooperate easily